spring-5-webflux/src/test/java/com/baeldung/spring/retry/ExternalConnectorIntegrationTest.java

```diff
@@ -0,0 +1,100 @@
+
+import static io.netty.handler.codec.http.HttpResponseStatus.SERVICE_UNAVAILABLE;
+import static io.netty.handler.codec.http.HttpResponseStatus.UNAUTHORIZED;
+import static org.assertj.core.api.Assertions.assertThat;
+
+class ExternalConnectorIntegrationTest {
+
+    private ExternalConnector externalConnector;
+
+    private MockWebServer mockExternalService;
+
+    @BeforeEach
+    void setup() throws IOException {
+       externalConnector = new ExternalConnector(
+               WebClient.builder().baseUrl("http://localhost:8090").build()
+       );
+       mockExternalService = new MockWebServer();
+       mockExternalService.start(8090);
+    }
+
+    @AfterEach
+    void tearDown() throws IOException {
+        mockExternalService.shutdown();
+    }
+
+    @Test
+    void givenExternalServiceReturnsError_whenGettingData_thenRetryAndReturnResponse() {
+
+        mockExternalService.enqueue(new MockResponse().setResponseCode(SERVICE_UNAVAILABLE.code()));
+        mockExternalService.enqueue(new MockResponse().setResponseCode(SERVICE_UNAVAILABLE.code()));
+        mockExternalService.enqueue(new MockResponse().setResponseCode(SERVICE_UNAVAILABLE.code()));
+        mockExternalService.enqueue(new MockResponse().setBody("stock data"));
+
+        StepVerifier.create(externalConnector.getData("ABC"))
+                .expectNextMatches(response -> response.equals("stock data"))
+                .verifyComplete();
+
+        verifyNumberOfGetRequests(4);
+    }
+
+    @Test
+    void givenExternalServiceReturnsClientError_whenGettingData_thenNoRetry() {
+
+        mockExternalService.enqueue(new MockResponse().setResponseCode(UNAUTHORIZED.code()));
+
+        StepVerifier.create(externalConnector.getData("ABC"))
+                .expectError(WebClientResponseException.class)
+                .verify();
+
+        verifyNumberOfGetRequests(1);
+    }
+
+    @Test
+    void givenExternalServiceRetryAttemptsExhausted_whenGettingData_thenRetryAndReturnError() {
+
+        mockExternalService.enqueue(new MockResponse().setResponseCode(SERVICE_UNAVAILABLE.code()));
+        mockExternalService.enqueue(new MockResponse().setResponseCode(SERVICE_UNAVAILABLE.code()));
+        mockExternalService.enqueue(new MockResponse().setResponseCode(SERVICE_UNAVAILABLE.code()));
+        mockExternalService.enqueue(new MockResponse().setResponseCode(SERVICE_UNAVAILABLE.code()));
+
+        StepVerifier.create(externalConnector.getData("ABC"))
+                .expectError(ServiceException.class)
+                .verify();
+
+        verifyNumberOfGetRequests(4);
+    }
+
+    private void verifyNumberOfGetRequests(int times) {
+        IntStream.range(0, times).forEach(i -> {
+            RecordedRequest recordedRequest = getRecordedRequest();
+            assertThat(recordedRequest.getMethod()).isEqualTo("GET");
+            assertThat(recordedRequest.getPath()).isEqualTo("/data/ABC");
+        });
+    }
+
+    private RecordedRequest getRecordedRequest() {
```

Comment:
This would be a lot cleaner if it just did a `throws Exception`, and if the `verifyNumberOfGetRequests` used a `for` loop and also allows `throws Exception`. While `IntStream` is nice for loops, it's not really much cleaner. Similarly, catching exceptions in tests, rather than propagating `throws` is a bit hard to reason about.

Reply:
changed to use ```for``` loop and ```exception``` clean up